### ISSUE TITLE
AddGrpcClient / CreateGrpcService lifetime

### ISSUE THREAD

**Issue by jamiewinder** [protobuf-net/protobuf-net.Grpc#11]
A couple of questions if that's okay:

- Is there an equivalent of `AddGrpcClient` that works with this library?
- Is the result of `CreateGrpcService` designed to be long-lived? e.g. inject a `HttpClient` into a singleton service, call `CreateGrpcService` on it once, and use this for the lifetime of the ASP.NET app. Is this safe?

**Comment by mgravell** [comment]
1. You'll have to point me at AddGrpcClient

2. Ultimately that comes down to the very complex story of HttpClient
lifetimes. I'd probably look first at the http-client-factory API and have
short-lived actual client instances.

**Comment by jamiewinder** [comment]
Thanks. Here is the `AddGrpcClient` methods I referred to:
https://github.com/grpc/grpc-dotnet/blob/e85b685f8965c7ac9b0145c90b0393a7f9201de1/src/Grpc.Net.ClientFactory/GrpcClientServiceExtensions.cs#L32

It's similar to `AddHttpClient`, it just creates a GRPC client of the specified contract instead.

**Comment by JamesNK** [comment]
@mgravell You should be able to reuse AddGrpcClient methods. Creating the client is self contained so you should just need to implement your own `GrpcClientFactory`

Default example here - https://github.com/grpc/grpc-dotnet/blob/e85b685f8965c7ac9b0145c90b0393a7f9201de1/src/Grpc.Net.ClientFactory/Internal/DefaultGrpcClientFactory.cs

**Comment by mgravell** [comment]
Thanks both, I'll take a look

**Comment by DJBMASTER** [comment]
I'm also looking into how to manage the lifetime of a GRPC Service through ASP.NET Core. 

Is it enough to keep a singleton GrpcChannel and call the various contract endpoints using that channel, or do singleton channels suffer from the same drawbacks as maintaining singleton HttpClients - rather than utilising IHTTPClientFactory?

**Comment by jamiewinder** [comment]
It looks like I've become a little out-of-sync with this project now. I'm not sure how I'm supposed to use a GRPC client from an ASP.NET Core Service? Up to now, I've just been injecting a `HttpClient` into my controllers / other services and using `.CreateGrpcService` on this to create the GRPC client. However, this method now appears to be gone (or rather, it now seems to be a method on `GrpcClientFactory`)?

**Comment by mgravell** [comment]
> Up to now, I've just been injecting a `HttpClient` into my controllers / other services and using `.CreateGrpcService` on this to create the GRPC client. 

shortly before GA, the MS bits changed the root type from `HttpClient` to `GrpcChannel` (which deals with the `HttpClient` internally).

To copy an example from [getting started](https://protobuf-net.github.io/protobuf-net.Grpc/gettingstarted):

``` c#
    using (var channel = GrpcChannel.ForAddress("http://localhost:10042"))
    {
        var calculator = channel.CreateGrpcService<ICalculator>();
        var result = await calculator.MultiplyAsync(new MultiplyRequest { X = 12, Y = 4 });
        Console.WriteLine(result.Result);
    }
```

Any problems: let me know.

**Comment by jamiewinder** [comment]
Thanks. So in the situation where an ASP.NET Service wants to communicate with a remote GRPC service, I should be creating and injecting a `GrpcChannel`, presumably as a singleton? Perhaps this is more a question of ASP.NET Core, but I think it'd be handy to see more about how this library can be used in  a microservice scenario where GRPC and ASP.NET Core really shine.

**Comment by mgravell** [comment]
Ultimately, the lifetime and usage of `GrpcChannel` is [*whatever asp.net says it is*](https://docs.microsoft.com/en-us/aspnet/core/grpc/client?view=aspnetcore-3.0); the only difference with protobuf-net.Grpc is that instead of using `new MyCalculator.CalculatorClient(channel)` you use `channel.CreateGrpcService<ICalculator>()`. I'm hesitant to offer usage guidance on `GrpcChannel` because *that isn't mine*.

**Comment by jamiewinder** [comment]
Thanks. I can appreciate that. I think the issue I'm having is - going back to the original intention for this issue - that `AddGrpcClient` exists in plain ASP.NET Core 3 which effectively hides / manages the creation and lifetimes of channels when using GRPC services, but I don't think an equivalent exists when using this library. Perhaps something just hasn't clicked for me yet, so I'll look into this again from the ASP.NET Core perspective.

Thanks again.

**Comment by mgravell** [comment]
You're right; I have not yet had opportunity to look at the `AddGrpcClient` API, or even investigate it. I don't have infinite time, and right now my priority is getting protobuf-net v3 shipped (which ties into reduced allocations for protobuf-net.Grpc).

If you can point to documentation (rather than code) on how that API is intended to use, it'll probably help me get started on it sooner...

**Comment by searus** [comment]
I think this is what @jamiewinder  is referring to: https://docs.microsoft.com/en-gb/aspnet/core/grpc/clientfactory?view=aspnetcore-3.1

I'd also be interested in understanding how to achieve the same with protobuf-net.Grpc

**Comment by mgravell** [comment]
I've finally had time to look into this, and essentially: the existing `ClientFactory` API is almost impossible for us to use directly, and even if we did - we'd have to duplicate most of the code; I have proposed some non-breaking API changes that would enable it - see https://github.com/grpc/grpc-dotnet/pull/837 (/cc @JamesNK )

**Comment by Reza-Noei** [comment]
@mgravell 

Any progress on this issue ?

**Comment by soroshsabz** [comment]
@mgravell as you know, https://github.com/grpc/grpc-dotnet/pull/851 has been merged, so I think this issue can continue ;)

**Comment by mgravell** [comment]
I'll need to refresh my understanding of this and have a look

**Comment by jhartmann123** [comment]
> Is there an equivalent of AddGrpcClient that works with this library?

As you can specify the factory for the GrpcService now, its as simple as that:
```cs
services.AddGrpcClient<ISomeService>(o =>
{
    o.Address = "https://some.service.url";
    o.Creator = ci => ci.CreateGrpcService<ISomeService>();
})
```

I've also written a small extension method for that:
```cs
public static IHttpClientBuilder AddCodeFirstGrpcClient<T>(this IServiceCollection services, Action<GrpcClientFactoryOptions> configureClient)
    where T : class
    => services.AddGrpcClient<T>(cfg =>
    {
        cfg.Creator = callInvoker => callInvoker.CreateGrpcService<T>();
        configureClient(cfg);
    });

// Call example:
services.AddCodeFirstGrpcClient<ISomeService>(o => o.Address = "https://some.service.url");
```

**Comment by soroshsabz** [comment]
@mgravell 

I am sorry for request again, but we need to this feature

thanks :)

**Comment by soroshsabz** [comment]
@jhartmann123 I am little confuse, what is difference your code with [protobuf-net.Grpc.ClientFactory ](https://www.nuget.org/packages/protobuf-net.Grpc.ClientFactory)?

The code exists in https://github.com/protobuf-net/protobuf-net.Grpc/blob/main/src/protobuf-net.Grpc.ClientFactory/ServicesExtensions.cs

**Comment by jhartmann123** [comment]
On the first glance, I also don't see any difference. I just didn't know about that lib. But I can't find any documentation about it either (docs or samples), am I missing something?

**Comment by soroshsabz** [comment]
@jhartmann123 So, if you have time to test this library, please share your feedback with us

thanks
